lib/netext/grpcext/conn.go

```diff
@@ -0,0 +1,313 @@
+// Package grpcext allows gRPC requests collecting stats info.
+package grpcext
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"go.k6.io/k6/js/modules"
+	"go.k6.io/k6/metrics"
+
+	protov1 "github.com/golang/protobuf/proto" //nolint:staticcheck,nolintlint // this is the old v1 version
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	grpcstats "google.golang.org/grpc/stats"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+// Request represents a gRPC request.
+type Request struct {
+	MethodDescriptor protoreflect.MethodDescriptor
+	Tags             map[string]string
+	Message          []byte
+}
+
+// Response represents a gRPC response.
+type Response struct {
+	Message  interface{}
+	Error    interface{}
+	Headers  map[string][]string
+	Trailers map[string][]string
+	Status   codes.Code
+}
+
+type clientConnCloser interface {
+	grpc.ClientConnInterface
+	Close() error
+}
+
+// Conn is a gRPC client connection.
+type Conn struct {
+	raw clientConnCloser
+}
+
+// DefaultOptions generates an option set
+// with common options for requests from a VU.
+func DefaultOptions(vu modules.VU) []grpc.DialOption {
+	dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+		return vu.State().Dialer.DialContext(ctx, "tcp", addr)
+	}
+
+	return []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.FailOnNonTempDialError(true),
+		grpc.WithReturnConnectionError(),
+		grpc.WithStatsHandler(statsHandler{vu: vu}),
+		grpc.WithContextDialer(dialer),
+	}
+}
+
+// Dial establish a gRPC connection.
+func Dial(ctx context.Context, addr string, options ...grpc.DialOption) (*Conn, error) {
+	conn, err := grpc.DialContext(ctx, addr, options...)
+	if err != nil {
+		return nil, err
+	}
+	return &Conn{
+		raw: conn,
+	}, nil
```

Comment:
how about including options in this struct so we don't have to pass options again for Invoke?

Reply:
I think [WithDefaultCallOptions](https://pkg.go.dev/google.golang.org/grpc#WithDefaultCallOptions) covers this case. Could it work for you?

Reply:
I'm a bit confused that options ...grpc.DialOption is passed in from `Dial` and `Invoke`. If options is part of the Conn struct, `Invoke` can use it as a receiver function.

Reply:
Would you add a slice DialOption field?
```go
type Conn struct {
        Options []grpc.DialOption

	raw clientConnCloser
}
```
and calling the `Conn.Invoke` method, would you like to not pass the Call Options?

```go
options := []grpc.DialOption{
   .....
}
conn, err := Dial(ctx, addr, options...)
err = conn.Invoke(ctx, url, md)
```

If yes, then the issue is mainly related to the different types involved. The eventual `Conn.Options` couldn't be used on the underhood call `grpc.Invoke` because the Invoke method wants the CallOption type.

Please, correct me if I misunderstood.

Reply:
I guess a User can just use https://pkg.go.dev/google.golang.org/grpc#WithDefaultCallOptions as well

Reply:
I see. They are different types of options. That makes sense.